### PR TITLE
Add Attempts to Job Persistence

### PR DESCRIPTION
**Pull request by cgardens**
## What
* We have a handful of problems that are caused by the fact that our worker implicitly has a concept of attempts but it is not explicitly encoded in the data models / persistence. Problems:
    * We lose information about attempts because they overwritten in the jobs database table.
    * The UI can't properly show logs. As new attempts start logs get overwritten.
    * The UI has to indefinitely poll for logs because it never knows when a job will start a new attempt. 
    * It is impossible to cancel a job from the UI, because the data model has no way of knowing it has been cancelled and shouldn't retry.

## How
* Add an attempts concept.
* This PR stops short of exposing it in the API. Going to do that in a separate smaller PR. The reason for this is that won't be merge-able until the UI is updated, so I want the smallest code footprint possible that is stuck in a "feature branch".
* Split the existing SchedulerPersistence into two interfaces. This split makes testing a lot easier and focuses both classes. 
    * JobPersistence - handles direct interaction with the database and governs the lifecycle rules jobs and attempts.
    * JobCreator - handles the conversion from the config data model into json objects that the JobPersistence can consume.

## Recommended reading order
1. `JobPersistence.java`
1. `DefaultJobPersistence.java`
1. `JobCreator.java`
1. `DefaultJobCreator.java`
1. `JobSubmitter.java`
1. the rest

## Follow up
* Instead of investing more in this, we might want to consider using an existing solution (e.g. airflow) in the future. Likely worth figuring out at the same time we figure out new destination abstraction.

### CODE REVIEW

File: airbyte-api/src/main/openapi/config.yaml [hunk 1]
```diff
@@ -1527,8 +1527,8 @@ components:
         - status
       properties:
         status:
-          $ref: "#/components/schemas/JobStatus"
-    JobStatus:
+          $ref: "#/components/schemas/JobStatusReadStatus"
+    JobStatusReadStatus:
```
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:1531]: this name is a mouthful, but needs to not collide with the more frequently used JobStatus

File: airbyte-db/src/main/java/io/airbyte/db/ExceptionWrappingDatabase.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+package io.airbyte.db;
+
+import java.io.IOException;
+import java.sql.SQLException;
+
+/**
+ * Wraps a {@link Database} object and throwing IOExceptions instead of SQLExceptions.
+ */
+public class ExceptionWrappingDatabase implements AutoCloseable {
```
Comment (cgardens) [inline airbyte-db/src/main/java/io/airbyte/db/ExceptionWrappingDatabase.java:33]: this saves lines and lines of try-catches in `DefaultJobPersistence`.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    final String scope = ScopeHelper.createScope(JobConfig.ConfigType.SYNC, connectionId.toString());
+
+    final JobSyncConfig jobSyncConfig = new JobSyncConfig()
+        .withSourceConnection(source)
+        .withSourceDockerImage(sourceDockerImageName)
+        .withDestinationConnection(destination)
+        .withDestinationDockerImage(destinationDockerImageName)
+        .withStandardSync(standardSync);
+
+    // todo (cgardens) - this will not have the intended behavior if the last job failed. then the next
```
Comment (cgardens) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java:141]: issue: https://github.com/airbytehq/airbyte/issues/1154. will fix this as my next task.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobStatus.java [hunk 4]
```diff
@@ -35,6 +35,6 @@
   COMPLETED,
   CANCELLED;
 
-  public static Set<JobStatus> TERMINAL_STATUSES = Sets.newHashSet(FAILED, COMPLETED, CANCELLED);
```
Comment (cgardens) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobStatus.java:38]: failed isn't terminal. we can recover from it.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobSubmitter.java [hunk 5]
```diff
@@ -91,31 +91,48 @@ public void run() {
   @VisibleForTesting
   void submitJob(Job job) {
     final WorkerRun workerRun = workerRunFactory.create(job);
+    // todo (cgardens) - we need to know the attempt id starts before the lifecycle begins otherwise we
+    // can't clean up properly on failures. this hack
+    // assumes that the submitter is always run in a single threaded manner, so we know can figure out
+    // what the next attempt id should be. we should move away from this hack. one approach would be to
+    // switch to uuids.
+    final long attemptId = job.getAttempts().size();
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobSubmitter.java:99]: The way we mix up database id and list index as id here makes me a bit nervous.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  private JobPersistence jobPersistence;
+
+  public DefaultJobCreator(JobPersistence jobPersistence) {
+    this.jobPersistence = jobPersistence;
+  }
+
+  @Override
+  public long createSourceCheckConnectionJob(SourceConnection source, String dockerImageName) throws IOException {
+    final String scope =
+        ScopeHelper.createScope(
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java:60]: What is required already to get rid of this scope concept?

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+          now);
+    })
+        .stream()
+        .findFirst()
+        .map(r -> r.get("attempt_id", Long.class))
+        .orElseThrow(() -> new RuntimeException("This should not happen"));
+  }
+
+  @Override
+  public void completeAttemptFailed(long jobId, long attemptId) throws IOException {
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:189]: NIT: `failAttempt`

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+          AttemptStatus.FAILED.toString().toLowerCase(),
+          now,
+          jobId,
+          attemptId);
+      return null;
+    });
+  }
+
+  @Override
+  public void completeAttemptSuccess(long jobId, long attemptId) throws IOException {
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:206]: `completeAttempt`

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/JobPersistence.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+  /**
+   * Set job status from current status to CANCELLED. If already in a terminal status, no op.
+   *
+   * @param jobId job to cancel
+   * @param cancellationReason rest that the job is being cancelled.
+   * @throws IOException exception due to interaction with persistence
+   */
+  void cancelJob(long jobId, CancellationReason cancellationReason) throws IOException;
+
+  enum CancellationReason {
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/JobPersistence.java:72]: I am really not a fan of the definition we are giving to cancellation, we are mixing failures & actual cancels

Comment (michel-tricot) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/JobPersistence.java:72]: Let's discuss it "in person"

File: airbyte-server/src/main/java/io/airbyte/server/handlers/JobHistoryHandler.java [hunk 10]
```diff
@@ -64,35 +71,56 @@ public JobReadList listJobsFor(JobListRequestBody request) throws IOException {
...
   public JobInfoRead getJobInfo(JobIdRequestBody jobIdRequestBody) throws IOException {
-    final Job job = schedulerPersistence.getJob(jobIdRequestBody.getId());
-
-    final LogRead logRead = new LogRead().logLines(IOs.getTail(LOG_TAIL_SIZE, job.getLogPath()));
+    final Job job = jobPersistence.getJob(jobIdRequestBody.getId());
 
     return new JobInfoRead()
         .job(getJobRead(job))
-        .logs(logRead);
+        .logs(job.getAttempts().stream().map(JobHistoryHandler::getAttemptInfoRead).map(AttemptInfoRead::getLogs).findFirst().orElse(null));
```
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/JobHistoryHandler.java:78]: NIT: can you go to a new line when you chain stream?

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java [hunk 11]
```diff
@@ -234,18 +241,18 @@ public JobStatusRead syncConnection(final ConnectionIdRequestBody connectionIdRe
         .build());
 
     return new JobStatusRead()
-        .status(job.getStatus().equals(JobStatus.COMPLETED) ? io.airbyte.api.model.JobStatus.SUCCEEDED : io.airbyte.api.model.JobStatus.FAILED);
+        .status(job.getStatus().equals(JobStatus.COMPLETED) ? JobStatusReadStatus.SUCCEEDED : JobStatusReadStatus.FAILED);
```
Comment (michel-tricot) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:244]: I honestly prefer the conflict than this name. By suffixing it with "ReadStatus" you prevent using this enum in other api models.

**Review by michel-tricot — CHANGES_REQUESTED** [review]

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/JobPersistence.java (hunk 9 above)
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/JobPersistence.java:72]: "in person" 🤣

**Comment by cgardens** [comment]
Per discussion, new state machine: 
Screen Shot 2020-12-02 at 9 36 59 AM

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobSubmitter.java (hunk 5 above)
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobSubmitter.java:99]: => attempt_number

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java (hunk 6 above)
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java:60]: the goal of the scope is to be able to group jobs that are created from the same config. so really any other solution that does that works fine.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java (hunk 11 above)
Comment (cgardens) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:244]: okay. i will try to address this in the API PR.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobStatus.java (hunk 4 above)
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobStatus.java:38]: fixed in new representation.

File: airbyte-db/src/main/resources/schema.sql [hunk 12]
```diff
@@ -32,15 +45,27 @@ CREATE
...
             created_at TIMESTAMPTZ,
             updated_at TIMESTAMPTZ
         );
 
+CREATE
+    TABLE
+        ATTEMPTS(
+            id BIGSERIAL PRIMARY KEY,
+            job_id BIGSERIAL,
+            attempt_id BIGSERIAL,
```
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-db/src/main/resources/schema.sql]: => attempt_name

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java [hunk 13]
```diff
@@ -85,9 +86,17 @@ private boolean shouldRetry(Job job) {
     return lastRun < timeSupplier.get().getEpochSecond() - TimeUnit.MINUTES.toSeconds(RETRY_WAIT_MINUTES);
   }
 
-  private void setSetStatusTo(Job job, JobStatus status) {
+  private void cancelJob(Job job) {
```
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java]: fixed in new state machine.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java (hunk 6 above)
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobCreator.java:60]: i haven't through of one yet that doesn't really closely resemble what we already have or that is much better. admittedly haven't spent too much time on this yet. this hasn't been a pain point at all. you have a specific looming concern?

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/JobPersistence.java (hunk 9 above)
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/JobPersistence.java]: fixed in new state machine.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java (hunk 11 above)
Comment (cgardens) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java]: basically i will see if i can get rid of this completely and just replace it with the existing JobRead struct. not really a good reason we have these separate structs right now.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java [hunk 14]
```diff
@@ -55,12 +55,12 @@ public void run() {
     listFailedJobs()
         .forEach(job -> {
           if (shouldCancel(job)) {
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java:57]: NIT: shouldCancel => hasReachMaxAttempt

**Review by michel-tricot — APPROVED** [review]
Amazing

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/Job.java [hunk 15]
```diff
@@ -105,6 +97,33 @@ public long getUpdatedAtInSecond() {
...
+      return Optional.empty();
+    }
+  }
+
+  public Optional<JobOutput> getSuccessOutput() {
+    return getSuccessfulAttempt().flatMap(Attempt::getOutput);
+  }
+
+  public boolean hasRunningAttempt() {
+    return getAttempts().stream().anyMatch(Attempt::isAttemptInTerminalState);
```
[the remark this answers is not in the document]
Comment (cgardens) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/Job.java:120]: fixed and added a test to catch this.

**Pull request merged**